### ISSUE TITLE
TableReporter<Vec3> requires that Vector_<Vec3> can be transposed to RowVector_<Vec3>

### ISSUE THREAD

**Issue by chrisdembia** [opensim-org/opensim-core#849]
@klshrinidhi this is the issue that we talked about; and I just ran into it with @aseth1's TableReporter.

@sherm1 do you know how to properly transpose a Vector_<Vec3> (or a RowVector_<Vec3>)? The transpose operator (~v) gives a compiler error: 

```
/Users/chris/repos/opensim/10/opensim-core/OpenSim/Common/Reporter.h:153:74: No viable conversion from 'THerm' (aka 'RowVectorView_<typename CNT<Vec<3, double, 1> >::THerm>') to 'const RowVector' (aka 'const RowVector_<SimTK::Vec<3, double, 1> >')
```

**Comment by sherm1** [comment]
That is the correct way to transpose a [Row]Vector. The problem appears to be in what gets done next with the result, which looks like an attempt to cast a fixed-size RowVectorView to a variable-size RowVector. More context would be helpful.

**Comment by chrisdembia** [comment]
Sure. Here's a small program that's trying to achieve the same thing:

```cpp
#include <Simbody.h>

using namespace SimTK;

int main() {
    Vector_<Vec3> v(5);

    RowVector_<Vec3> rv(~v);
}
```

And here is the compiler error:

```cpp
/Users/chris/repos/simbody/4/install/share/doc/simbody/examples/src/vectorvec3transpose.cpp:8:22: error: 
      no matching constructor for initialization of 'RowVector_<Vec3>'
    RowVector_<Vec3> rv(~v);
                     ^  ~~
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:65:5: note: 
      candidate constructor not viable: no known conversion from 'THerm' (aka
      'RowVectorView_<typename CNT<Vec<3, double, 1> >::THerm>') to 'const
      SimTK::RowVector_<SimTK::Vec<3, double, 1> >' for 1st argument
    RowVector_(const RowVector_& src) : Base(src) {}
    ^
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:68:5: note: 
      candidate constructor not viable: no known conversion from 'THerm' (aka
      'RowVectorView_<typename CNT<Vec<3, double, 1> >::THerm>') to 'const Base'
      (aka 'const RowVectorBase<SimTK::Vec<3, double, 1> >') for 1st argument
    RowVector_(const Base& src) : Base(src) {}    // e.g., RowVectorView
    ^
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:69:5: note: 
      candidate constructor not viable: no known conversion from 'THerm' (aka
      'RowVectorView_<typename CNT<Vec<3, double, 1> >::THerm>') to
      'const BaseNeg' (aka 'const RowVectorBase<ENeg>') for 1st argument
    RowVector_(const BaseNeg& src) : Base(src) {}  
    ^
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:78:14: note: 
      candidate constructor not viable: no known conversion from 'THerm' (aka
      'RowVectorView_<typename CNT<Vec<3, double, 1> >::THerm>') to 'int' for
      1st argument
    explicit RowVector_(int n) : Base(n) { }
             ^
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:97:14: note: 
      candidate template ignored: could not match 'Row' against 'RowVectorView_'
    explicit RowVector_(const Row<M,ELT>& v) : Base(M) {
             ^
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:61:5: note: 
      candidate constructor not viable: requires 0 arguments, but 1 was provided
    RowVector_() : Base() {}   // 1x0 reallocatable
    ^
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:79:5: note: 
      candidate constructor not viable: requires 2 arguments, but 1 was provided
    RowVector_(int n, const ELT* cppInitialValues) : Base(n, cppInitialV...
    ^
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:80:5: note: 
      candidate constructor not viable: requires 2 arguments, but 1 was provided
    RowVector_(int n, const ELT& initialValue)     : Base(n, initialValue) {}
    ^
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:86:5: note: 
      candidate constructor not viable: requires 3 arguments, but 1 was provided
    RowVector_(int n, const S* cppData, bool): Base(n, Base::CppNScalars...
    ^
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:87:5: note: 
      candidate constructor not viable: requires 3 arguments, but 1 was provided
    RowVector_(int n,       S* cppData, bool): Base(n, Base::CppNScalars...
    ^
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:92:5: note: 
      candidate constructor not viable: requires 4 arguments, but 1 was provided
    RowVector_(int n, int stride, const S* data, bool) : Base(n, stride...
    ^
/Users/chris/repos/simbody/4/install/include/simbody/SimTKcommon/internal/RowVector_.h:93:5: note: 
      candidate constructor not viable: requires 4 arguments, but 1 was provided
    RowVector_(int n, int stride,       S* data, bool) : Base(n, stride...
    ^
```

**Comment by chrisdembia** [comment]
Based on your earlier comment, I just tried a small change to the program above and now it compiles! Is this the right fix?

```cpp
#include <Simbody.h>

using namespace SimTK;

int main() {
    Vector_<Vec3> v(5);

    RowVector_<Vec3> rv(v.size());
    
    rv = ~v;
}
```

**Comment by chrisdembia** [comment]
Hmm...maybe not... if I add `std::cout << rv << std::endl;`, I get this exception:

```
libc++abi.dylib: terminating with uncaught exception of type SimTK::Exception::Cant: SimTK Exception thrown at MatrixHelperRep.h:601:
  Can't perform operation: One-index getElt_() not available for 2D matrices
```

**Comment by chrisdembia** [comment]
Okay well this works:

```
RowVector_<Vec3> rv(v.size(), &v[0]);
```

**Comment by chrisdembia** [comment]
As a note, the original code works fine if `T = double` instead of `T = Vec3`.

**Comment by sherm1** [comment]
This probably indicates some constructors, conversions, and methods are missing from the RowVector class. I think it is less complete than Vector because it doesn't get much use. They are probably very simple things to add -- please file a Simbody issue so we don't lose track (PR would be good too if you're feeling brave!). Your code:
```cpp
#include <Simbody.h>
using namespace SimTK;

int main() {
    Vector_<Vec3> v(5);
    RowVector_<Vec3> rv(v.size());   
    rv = ~v;
}
```
is correct. I think the error you got when writing it out was due to a missing RowVector method causing it to attempt to use the base class Matrix for something that should have been overridden but wasn't (a bug).

This code:
```cpp
RowVector_<Vec3> rv(v.size(), &v[0]);
```
is using the space-sharing constructor rather than copying the data as the previous code did.

**Comment by chrisdembia** [comment]
> please file a Simbody issue so we don't lose track

Done.

> (PR would be good too if you're feeling brave!)

Not feeling brave today; feeling more like :chicken:.
> This code:
> 
> RowVector_<Vec3> rv(v.size(), &v[0]);
> is using the space-sharing constructor rather than copying the data as the previous code did.

Are you sure? Wouldn't space-sharing (borrowing?) use this constructor: `rv(v.size(), &v[0], true)`?

**Comment by sherm1** [comment]
> Are you sure? Wouldn't space-sharing (borrowing?) use this constructor: rv(v.size(), &v[0], true)?

You're right! Sorry, I forgot about that last argument.

**Comment by chrisdembia** [comment]
Thank you @sherm1 for diagnosing this.

**Comment by jenhicks** [comment]
@klshrinidhi will investigate what can be done in C++ and propose a solution for scripting.

**Comment by sherm1** [comment]
@klshrinidhi, I think the transpose of a `Vector_<Vec3>` is actually a `RowVector_<Row3>` (that is, each of the elements is transposed too). That is usually the right semantics for transpose - for example, if the elements are complex then the transpose should have complex conjugate elements (e.g. [here](https://en.wikipedia.org/wiki/Conjugate_transpose)). Simbody has a notion of "positional transpose" for the case where you want to transpose only the container but not the elements.

I'm just mentioning this in case this has something to do with the problem encountered here.

**Comment by klshrinidhi** [comment]
Solution for C++ is mentioned in comments in this issue.
Solution for Python/Java is merged by PR #1470.
